### DISCUSSION TITLE
How to use citext in module?

### DISCUSSION THREAD

**Discussion by Sleepful** [diesel-rs/diesel#4046] [Q&A] [answered]
```rs
use diesel::prelude::*;
use std::time::SystemTime;


#[derive(Queryable, Selectable)]
#[diesel(table_name = crate::schema::users)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct User {
    pub id: i64,
    pub email: String,
}
```
This isn't working with `String` because I get this error:

```
src/models.rs|16 col 16-22 error| the trait bound `*const str: FromSql<Citext, Pg>` is not satisfied the trait `FromSql<diesel::sql_types::Text, Pg>` is implemented for `*const str` for that trait implementation, expected `diesel::sql_types::Text`, found `Citext` required for `String` to implement `FromSql<Citext, Pg>` required for `String` to implement `diesel::Queryable<Citext, Pg>` required for `String` to implement `FromSqlRow<Citext, Pg>` see issue #48214
```

I am on `{ version = "2.1.6", features = ["postgres"] }`

I did notice in the docs that CITEXT doesn't appear for 2.1.6:
- https://docs.diesel.rs/2.1.x/diesel/sql_types/index.html
It does appear for master (which indicates 2.1.4 in top of sidebar):
- https://docs.diesel.rs/master/diesel/sql_types/index.html

I also noticed that the generated `schema.rs` includes this:

```rs
pub mod sql_types {
    #[derive(diesel::query_builder::QueryId, diesel::sql_types::SqlType)]
    #[diesel(postgres_type(name = "citext"))]
    pub struct Citext;
}
```
What should I use as a type on my model for the Citext column?

**Comment by weiznich** [comment, marked as answer]
As you already noticed: The `citext` type is not supported on any released diesel version yet. Support for it has landed on the master branch and will be included in the next feature release. Until then you need to provide your own type mapping for the generated type as you would do for any other custom type.

**Reply by Sleepful** [reply]
OH thank you, I am still wrapping my mind around things here.

Taking from [the code](https://docs.diesel.rs/master/src/diesel/pg/types/primitives.rs.html#83-105) that has not been included for the current release, this satisfies the compiler:

```rs
use crate::schema::sql_types::Citext;
use diesel::deserialize::{self, FromSql};
use diesel::pg::{Pg, PgValue};
use diesel::serialize::{self, IsNull, Output, ToSql};
use std::io::Write;

impl ToSql<Citext, Pg> for String {
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Pg>) -> serialize::Result {
        out.write_all(self.as_bytes())?;
        Ok(IsNull::No)
    }
}
impl FromSql<Citext, Pg> for String {
    fn from_sql(value: PgValue<'_>) -> deserialize::Result<Self> {
        let string = String::from_utf8(value.as_bytes().to_vec())?;
        Ok(string)
    }
}
```
